chapter/4/dist-langs.md

```diff
@@ -4,7 +4,515 @@ title:  "Distributed Programming Languages"
+Units of distribution in Erlang are processes.
+These processes may be co-located on the same node or distributed amongst a cluster.
+
+Processes in Erlang communicate by message passing.
+Specifically, a process can send an asynchronous message to another process' *mailbox*.
+At some future time, the receiving process may enter a *receive* clause, which searches the mailbox for the first message that matches one of a set of patterns.
+The branch of code that is executed in response to a message is dependent on the pattern that is matched.
+
+In general, an application written in Erlang is separated into two broad components: *workers* and *monitors*.
+Workers are responsible for application logic.
+Erlang offers a special function `link(Pid)` which allows one process to monitor another.
+If the process indicated by `Pid` fails, the monitor process will be notified and is expected to handle the error.
+Worker processes are "linked" by monitor processes which implement the fault-tolerance logic of the application.
+
+Erlang, first implemented in Prolog, has the features and styles of a functional programming language.
+Variables in Erlang are immutable; once assigned a value, they cannot be changed.
+Because of this, loops in Erlang are written as tail recursive function calls.
+Although this at first seems like a flawed practice (to traditional procedural programmers), tail recursive calls do not grow the current stack frame, but instead replace it.
+It is worth noting that stack frame growth is still possible, but if the recursive call is "alone" (the result of the inner function is the result of the outer function), the stack will not grow.
+
+#### Orleans (2011)
+
+Orleans is a programming model for distributed computing based on actors.
+An Orleans program can be conceptualized as a collection of actors.
+Because Orleans is intended as a model for building cloud applications, actors do not spawn independent processes as they do in Emerald or Argus.
+Rather, Orleans actors are designed to execute only when responding to requests.
+
+As in Argus, Orleans encapsulates requests (root function calls) within *transactions*.
+When processing a transaction, function calls may span many actors.
+To ensure consistency in the end result of a transaction, Orleans offers another abstraction, *activations*, to allow each transaction to operate on a consistent state of actors.
+An activation is an instance of an actor.
+Activations allow (1) consistent access to actors during concurrent transactions, and (2) high throughput when an actor becomes "hot."
+Consistency is achieved by only allowing transactions to "touch" activations of actors that are not being used by another transaction.
+High throughput is achieved by spawning many activations of an actor for handling concurrent requests.
+
+For example, suppose there is an actor that represents a specific YouTube video.
```

Comment:
These next two paragraphs are very disconnected and feel like they are floating.